Feature: Improve performance of the search page

jellebuitenhuisis opened the pull request:
I am just going to leave comments on what I've done in the relevant parts. I think that might be the easiest way to walk someone through this PR.

Closes #160

jellebuitenhuisis commented on src/app/find-the-expert/page.tsx:
```diff
@@ -37,14 +35,26 @@ const ContentSection = ({
     </>
 );
 
+const FindTheExpertSearch = async () => {
```
Originally an attempt was made to reuse code in `ShowRolesWrapper`. However this included weakly typed string comparison:
`if(path.includes("expert"))`

I've split this into two components, because they do different things. Originally ShowRolesWrapper made this distinction. Here it is made on the page where the search component is used itself

jellebuitenhuisis commented on src/app/find-the-expert/page.tsx:
```diff
@@ -74,14 +84,6 @@ const FindTheExpertPage = async (context: {
     );
 };
 
-async function getUserAnswers(tech?: string, role?: string, unit?: string) {
```
A function which only calls another function adds nothing of value

jellebuitenhuisis commented on src/app/find-the-expert/page.tsx:
```diff
@@ -91,19 +93,17 @@ const ShowTableWrapper = async ({
-        userIds,
-        answerIds,
-    );
+    const { dataByRoleAndQuestion, aggregatedDataByRole } =
```
Instead of multiple database calls, these have been merged into one function. This outputs the data which `ShowDataTable` needs.
`ShowDataTable` is just a component for displaying things, it should not also have to do lots of data manipulation, especially considering it's done client side.

jellebuitenhuisis commented on src/app/page.tsx:
```diff
@@ -1,5 +1,5 @@
 import { getServerAuthSession } from "~/server/auth";
-import SelectRole from "../components/select-input";
+import SelectRole from "~/components/select-input";
```
Always use absolute imports!

nicojs replied:
![](https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExZ3NkZ3NuNGNpcXk5bWJ1c3ZqcDViMWFkOWhlbXR4ZnYwa2xsM3ZxMiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/l0ExbnGIX9sMFS7PG/giphy.gif)

jellebuitenhuisis commented on src/app/result/page.tsx:
```diff
@@ -59,27 +58,14 @@ const Results = async (context: {
-    const def: Section = {
+    const def: Role = {
         id: "",
-        href: path,
```
Lots of unnecessary properties that weren't used at all. I've removed them

jellebuitenhuisis commented on src/app/result/page.tsx:
```diff
@@ -143,7 +129,7 @@ const FetchQuestionResults = async ({
     }
 
-    return await db.questionResult.findMany({
+    return db.questionResult.findMany({
```
It's already an async function, this await is redundant

jellebuitenhuisis commented on src/components/data-tables/show-data-table.tsx:
```diff
@@ -0,0 +1,144 @@
+"use client";
```
This file was moved and not newly created, but git doesn't understand that...

jellebuitenhuisis commented on src/components/data-tables/show-data-table.tsx:
```diff
@@ -0,0 +1,144 @@
+        Object.keys(aggregatedDataByRole)[0] ?? "",
+    ]);
+
+    const toggleRoleExpansion = (role: string) => {
```
On the first page load, we were showing 273(!) tables. This grinded the users page to a halt. I've chosen to initially only show one table, the first one. The user can toggle other tables by using the down arrow next to the table title.

jellebuitenhuisis commented on src/components/show-data-table.tsx:
```diff
@@ -1,161 +0,0 @@
-                </p>
-            ) : (
-                Object.keys(dataByRoleAndQuestion).map((role) => {
-                    if (
```
This comparison is unnecessary, we are already filtering based on role in the first where clause at the database query

jellebuitenhuisis commented on src/components/show-data-table.tsx:
```diff
@@ -1,161 +0,0 @@
-              )
-            : users;
-
-    const { userMap, answerOptionMap } = createUserAndAnswerMaps(
```
Because we are passing the data this component needs directly, all these calculations on the client side have now been removed

jellebuitenhuisis commented on src/components/ui/show-tech-search-wrapper.tsx:
```diff
@@ -90,15 +92,27 @@ const ShowTechSearchWrapper = ({
     }, [unit, logUsageMetric]);
 
+    // Create a debounced function to update the URL
+    const debouncedUpdateURL = useDebouncedCallback(
```
Don't immediately update the URL when the user types. See the [Next.js tutorial](https://nextjs.org/learn/dashboard-app/adding-search-and-pagination#best-practice-debouncing) for an example

jellebuitenhuisis commented on src/utils/client-data-manipulation.ts:
```diff
@@ -1,268 +0,0 @@
-"use client";
```
This file was deleted, because we are now doing all of this server side

jellebuitenhuisis commented on src/utils/role-utils.ts:
```diff
@@ -1,40 +1,18 @@
-            })
-            .map((role) => ({
-                id: role.id,
-                href: createHref(resultHref, role),
```
Lots of unused properties, so this file was rewritten

jellebuitenhuisis commented on src/utils/data-manipulation.ts:
```diff
@@ -1,60 +1,262 @@
+}) => {
+    const where = buildWhereClause({ role, questionText, unit });
+
+    const questionResults = await db.questionResult.findMany({
```
Originally we did 3 database queries here. First we got the questionresults with a where filter.
Using those question results, we extracted the unique user-/answer id's. Based on those id's, we went to the database twice more. Once to get users and once to find the answeroptions.

This can all be combined into one query. We still find the questionResults, but we include the user and answers that are connected to that question. This saves two round trips

jellebuitenhuisis commented on src/utils/data-manipulation.ts:
```diff
@@ -1,60 +1,262 @@
+};
+
+const groupDataByRoleAndQuestion = (
+    questionResults: QuestionResultWithRelations[],
```
This function was originally done client side. It was reworked to use the result of the database query directly, instead of relying on an intermediate userMap and answerOptionMap. The output is still the same

jellebuitenhuisis commented on src/utils/data-manipulation.ts:
```diff
@@ -1,60 +1,262 @@
+const getRoleName = (role: Role) => role.role || "Unknown Role";
+
+const aggregateDataByRole = (
+    questionResults: QuestionResultWithRelations[],
```
This function was originally done client side. It was reworked to use the result of the database query directly, instead of relying on an intermediate userMap and answerOptionMap. The output is still the same

nicojs commented on src/utils/data-manipulation.ts:
```diff
@@ -1,60 +1,262 @@
+    };
+}>;
+
+export const fetchAnswerData = async ({
```
I think this method name can be improved, something like `retrieveAnswersByRole`

nicojs reviewed (approved):
I've got some small remarks but otherwise ok.

So just to be clear, to fix the performance bottlenecks:
1. We've moved some stuff serverside
2. We've implemented debounce using `use-debounce` so search only happens when te user stops typing.

And we hope this is enough? Because we still load all the data, right?

jellebuitenhuisis commented:
> I've got some small remarks but otherwise ok.
> 
> So just to be clear, to fix the performance bottlenecks:
> 
> 1. We've moved some stuff serverside
> 2. We've implemented debounce using `use-debounce` so search only happens when te user stops typing.
> 
> And we hope this is enough? Because we still load all the data, right?

We also don't show all the data at once. If no role is selected, we only show the first table, the rest is collapsed and not rendered yet. Part of the performance problem was in the multiple database calls and subsequent client-side data processing. The other part was that the browser would turn slow, because showing 273 filled tables at the same time, is quite a lot.
